File: xform-js/src/ast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuleDef = exports.FunctionDef = exports.Param = exports.AttributePattern = exports.TypedPattern = exports.ElementPattern = exports.WildcardPattern = exports.StepTest = exports.PathStep = exports.PathStart = exports.Interp = exports.Text = exports.TextConstructor = exports.Constructor = exports.PathExpr = exports.BinaryOp = exports.UnaryOp = exports.FuncCall = exports.MatchExpr = exports.ForExpr = exports.LetExpr = exports.IfExpr = exports.VarRef = exports.Literal = exports.Module = void 0;
class Module {
    constructor(opts) {
        this.functions = opts.functions;
        this.rules = opts.rules;
        this.vars = opts.vars;
        this.namespaces = opts.namespaces;
        this.imports = opts.imports;
        this.expr = opts.expr;
    }
}
exports.Module = Module;
class Literal {
    constructor(value) {
        this.value = value;
    }
}
exports.Literal = Literal;
class VarRef {
    constructor(name) {
        this.name = name;
    }
}
exports.VarRef = VarRef;
class IfExpr {
    constructor(cond, thenExpr, elseExpr) {
        this.cond = cond;
        this.then_expr = thenExpr;
        this.else_expr = elseExpr;
    }
}
exports.IfExpr = IfExpr;
class LetExpr {
    constructor(name, value, body) {
        this.name = name;
        this.value = value;
        this.body = body;
    }
}
exports.LetExpr = LetExpr;
class ForExpr {
    constructor(name, seq, where, body) {
        this.name = name;
        this.seq = seq;
        this.where = where;
        this.body = body;
    }
}
exports.ForExpr = ForExpr;
class MatchExpr {
    constructor(target, cases, defaultExpr) {
        this.target = target;
        this.cases = cases;
        this.defaultExpr = defaultExpr;
    }
}
exports.MatchExpr = MatchExpr;
class FuncCall {
    constructor(name, args) {
        this.name = name;
        this.args = args;
    }
}
exports.FuncCall = FuncCall;
class UnaryOp {
    constructor(op, expr) {
        this.op = op;
        this.expr = expr;
    }
}
exports.UnaryOp = UnaryOp;
class BinaryOp {
    constructor(op, left, right) {
        this.op = op;
        this.left = left;
        this.right = right;
    }
}
exports.BinaryOp = BinaryOp;
class PathExpr {
    constructor(start, steps) {
        this.start = start;
        this.steps = steps;
    }
}
exports.PathExpr = PathExpr;
class Constructor {
    constructor(name, attrs, contents) {
        this.name = name;
        this.attrs = attrs;
        this.contents = contents;
    }
}
exports.Constructor = Constructor;
class TextConstructor {
    constructor(expr) {
        this.expr = expr;
    }
}
exports.TextConstructor = TextConstructor;
class Text {
    constructor(value) {
        this.value = value;
    }
}
exports.Text = Text;
class Interp {
    constructor(expr) {
        this.expr = expr;
    }
}
exports.Interp = Interp;
class PathStart {
    constructor(kind, name = null) {
        this.kind = kind;
        this.name = name;
    }
}
exports.PathStart = PathStart;
class PathStep {
    constructor(axis, test, predicates) {
        this.axis = axis;
        this.test = test;
        this.predicates = predicates;
    }
}
exports.PathStep = PathStep;
class StepTest {
    constructor(kind, name = null) {
        this.kind = kind;
        this.name = name;
    }
}
exports.StepTest = StepTest;
class WildcardPattern {
}
exports.WildcardPattern = WildcardPattern;
class ElementPattern {
    constructor(name, varName = null, child = null) {
        this.name = name;
        this.varName = varName;
        this.child = child;
    }
}
exports.ElementPattern = ElementPattern;
class TypedPattern {
    constructor(kind) {
        this.kind = kind;
    }
}
exports.TypedPattern = TypedPattern;
class AttributePattern {
    constructor(name) {
        this.name = name;
    }
}
exports.AttributePattern = AttributePattern;
class Param {
    constructor(name, typeRef = null, defaultExpr = null) {
        this.name = name;
        this.type_ref = typeRef;
        this.defaultExpr = defaultExpr;
    }
}
exports.Param = Param;
class FunctionDef {
    constructor(params, body) {
        this.params = params;
        this.body = body;
    }
}
exports.FunctionDef = FunctionDef;
class RuleDef {
    constructor(pattern, body) {
        this.pattern = pattern;
        this.body = body;
    }
}
exports.RuleDef = RuleDef;

File: xform-kotlin/src/main/kotlin/com/zopyx/xform/Ast.kt
package com.zopyx.xform

// AST Expressions
sealed interface Expr

data class Literal(val value: Any?) : Expr
data class VarRef(val name: String) : Expr
data class IfExpr(val cond: Expr, val thenExpr: Expr, val elseExpr: Expr) : Expr
data class LetExpr(val name: String, val value: Expr, val body: Expr) : Expr
data class ForExpr(val name: String, val seq: Expr, val where: Expr?, val body: Expr) : Expr
data class MatchExpr(val target: Expr, val cases: List<MatchCase>, val default: Expr?) : Expr
data class MatchCase(val pattern: Pattern, val expr: Expr)
data class FuncCall(val name: String, val args: List<Expr>) : Expr
data class UnaryOp(val op: String, val expr: Expr) : Expr
data class BinaryOp(val op: String, val left: Expr, val right: Expr) : Expr
data class PathExpr(val start: PathStart, val steps: List<PathStep>) : Expr
data class Constructor(val name: String, val attrs: List<AttrConstructor>, val contents: List<Expr>) : Expr
data class AttrConstructor(val name: String, val expr: Expr)
data class TextConstructor(val expr: Expr) : Expr
data class Text(val value: String) : Expr
data class Interp(val expr: Expr) : Expr

// Path expressions
data class PathStart(val kind: String, val name: String? = null)
data class PathStep(val axis: String, val test: StepTest, val predicates: List<Expr> = emptyList())
data class StepTest(val kind: String, val name: String? = null)

// Patterns
sealed interface Pattern
object WildcardPattern : Pattern
data class ElementPattern(val name: String, val variable: String? = null, val child: Pattern? = null) : Pattern
data class TypedPattern(val kind: String) : Pattern
data class AttributePattern(val name: String) : Pattern

// Module declarations
data class Module(
    val functions: MutableMap<String, FunctionDef> = mutableMapOf(),
    val rules: MutableMap<String, MutableList<RuleDef>> = mutableMapOf(),
    val vars: MutableMap<String, Expr> = mutableMapOf(),
    val namespaces: MutableMap<String, String> = mutableMapOf(),
    val imports: MutableList<ImportDecl> = mutableListOf(),
    var expr: Expr? = null
)

data class ImportDecl(val iri: String, val alias: String? = null)
data class Param(val name: String, val typeRef: String? = null, val default: Expr? = null)
data class FunctionDef(val params: List<Param>, val body: Expr)
data class RuleDef(val pattern: Pattern, val body: Expr)

// Function reference (for passing functions as arguments)
data class FunctionRef(val name: String)
